pkg/planner/core/rule_decorrelate.go

```diff
@@ -133,8 +133,71 @@ func (*DecorrelateSolver) aggDefaultValueMap(agg *logicalop.LogicalAggregation)
 	return defaultValueMap
 }
 
+// pruneRedundantApply: Removes the Apply operator if the parent SELECT clause does not filter any rows from the source.
+// Example: SELECT 1 FROM t1 AS tab WHERE 1 = 1 OR (EXISTS(SELECT 1 FROM t2 WHERE a2 = a1))
+// In this case, the subquery can be removed entirely since the WHERE clause always evaluates to True.
+// This results in a SELECT node with a True condition and an Apply operator as its child.
+// If this pattern is detected, we remove both the SELECT and Apply nodes, returning the left child of the Apply operator as the result.
+// For the example above, the result would be a table scan on t1.
+func pruneRedundantApply(p base.LogicalPlan) (base.LogicalPlan, bool) {
+	// Check if the current plan is a LogicalSelection
+	logicalSelection, ok := p.(*logicalop.LogicalSelection)
+	if !ok {
+		return nil, false
+	}
+
+	// Retrieve the child of LogicalSelection
+	selectSource := logicalSelection.Children()[0]
+
+	// Check if the child is a LogicalApply
+	apply, ok := selectSource.(*logicalop.LogicalApply)
+	if !ok {
+		return nil, false
+	}
+
+	// Ensure the Apply operator is of a suitable join type to match the required pattern.
+	// Only LeftOuterJoin or LeftOuterSemiJoin are considered valid here.
+	if apply.JoinType != logicalop.LeftOuterJoin && apply.JoinType != logicalop.LeftOuterSemiJoin {
```

Comment:
can we add one more line check here?
LeftOuter apply's schema.columns[len-1] (whichi is the scalar result output column) is in the selection's condition, which will make sure that the apply's scalar output is used just in the parent selection.

Reply:
I think we can add a test to check it.
It should the case like `select ((1 = 1) or (exists(...)) from t`. So there'll be a final projection to project the result because the extra column of the semi-join only indicates that the join result is matched or not.
Now that the semi-join's execution result is useless, the extra column should also be useless.

Reply:
The outer join should be sufficient since it only applies if you have subquery predicate within a complex condition. @winoros : did not get your test suggestion. Is the subquery in the select list or where cluase? If where clause then it is already covered. Also, as far as the extra projection, prune columns is applied again at the end of the rules which removes the subquery result column.